### PR TITLE
test: Add tests for logging

### PR DESCRIPTION
**Pull request by vvoland**
Supersedes: 
- https://github.com/moby/moby/pull/43642

(extended with: https://github.com/moby/moby/issues/43647#issuecomment-1141393871)

Add integration tests for the ContainerLogs API call
Each test handle a distinct case of ContainerLogs output.
- Muxed stream, when container is started without tty
- Single stream, when container is started with tty

Add unit test for LogReader suite that tests concurrent logging
It checks that there are no race conditions when logging concurrently
from multiple goroutines.

The tests pass for master.

With the https://github.com/moby/moby/pull/43650 reverted they fail as expected.
In such case, also `TestReadLog/Concurrent` should also fail when not running
with `-race` due to checking if the read messages match those that were logged

**- What I did**
Write unit test for race condition when logging from multiple goroutines at once

**- How to verify it**
1. Run unit tests:
```
make TESTDIRS='github.com/docker/docker/daemon/logger/local'  BIND_DIR=. test-unit
make TESTDIRS='github.com/docker/docker/daemon/logger/jsonfilelog'  BIND_DIR=. test-unit
```
or (is there any way to pass -race to the test-unit?)
```
cd daemon/logger/local
go test -race .
cd -
cd daemon/logger/jsonfilelog
go test -race .
```

2. Run integration tests:
```
make BIND_DIR=. TEST_FILTER='TestLogs' test-integration
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by corhere — APPROVED** [review]
Thanks for adding the concurrent race test! The PR looks good to me, though I have some nits and suggestions.

**Review by corhere — CHANGES_REQUESTED** [review]
Sorry; just thought of something: I would like to see both the `json-file` and `local` drivers covered by the integration tests. (Maybe one day `journald` could join that list, [but that's a whole 'nother PR.](https://github.com/moby/moby/pull/43219))

**Comment by vvoland** [comment]
Thanks @corhere. I applied your suggestion. Also I encountered some trouble with log rotation.

So, the test first logged the messages and then checked if the messages read from LogReader match messages that were logged. This made the TestTailLogsWithRotation test tree fail, because due to low `max-size` the log can't store all the messages, so only the last 10 messages could be read.
This made me think, that considering the nature of Tail, it is not reasonable to expect to store all the messages that were ever logged and it would probably be better for the test to use the Follow, because:
https://github.com/moby/moby/blob/3e32104e0e39af9019a3ea9aa7093bb7c97fcf05/daemon/logger/loggertest/logreader.go#L173-L177

So I changed the test to use Follow instead and it worked at first, but when I ran the test again a few times then it eventually failed:

```
time="2022-06-01T16:16:37+02:00" level=warning msg="file rotations were missed while following logs; some log messages have been skipped over" file="/var/folders/rv/7kmyxm9n6hd6xcpsh8tx3g300000gn/T/TestFollowLogsWithRotationcompress=falseConcurrent3106684161/001/logconcurrent0.log" missed-rotations=2 module=logger
time="2022-06-01T16:16:37+02:00" level=warning msg="error closing current log file" error="invalid argument" file="/var/folders/rv/7kmyxm9n6hd6xcpsh8tx3g300000gn/T/TestFollowLogsWithRotationcompress=falseConcurrent3106684161/001/logconcurrent0.log" module=logger
--- FAIL: TestFollowLogsWithRotation (0.00s)
    --- FAIL: TestFollowLogsWithRotation/compress=false (0.00s)
        --- FAIL: TestFollowLogsWithRotation/compress=false/Concurrent (0.02s)
            logreader.go:446: assertion failed: stdout msg 167 (string) != stdout msg 112 (string)
            read_test.go:174: TestFollowLogsWithRotation/compress=false/Concurrent:

                logconcurrent0.log      -rw-r-----      375B    2022-06-01 16:16:37.628204038 +0200 CEST
                logconcurrent0.log.1    -rw-r-----      4125B   2022-06-01 16:16:37.627307401 +0200 CEST
                logconcurrent0.log.2    -rw-r-----      4158B   2022-06-01 16:16:37.623293432 +0200 CEST
FAIL
FAIL    github.com/docker/docker/daemon/logger/jsonfilelog      1.207s
FAIL
```


So it seems like the LogFile is not able to guarantee that no messages are dropped when rotations happen.
This can be solved by adjusting the test logger configuration to make it perform at most one rotation:
https://github.com/moby/moby/blob/3e32104e0e39af9019a3ea9aa7093bb7c97fcf05/daemon/logger/jsonfilelog/read_test.go#L162-L170

So I increased it to make the test not exceed one rotation. This makes other tests from the LogReader suite a bit pointless under the WithRotation logger, because they won't even have a chance for any rotation to occur, due to the low amount of messages that they log. TestConcurrent logs a lot of messages (10000), to make it more possible for the race condition to be visible, when not running the test with `-race`. 

Wondering what's the best decision here. Some options:
1. Not execute TestConcurrent for TestTailLogsWithRotation
2. Only log the messages  concurrently checking for race condition without reading and checking if read messages are as expected. (Would it still be a LogReader test though? 💭) 
3. Make TestConcurrent use the same messages as other tests (<10 messages) - it will still catch the race condition if run under `go test -race`, but may fail to catch it otherwise. (our CI doesn't run tests with `-race` yet)

Ideally we would want the LogFile to not drop messages when rotation happen quickly, but it's probably not feasible.

3 sounds reasonable to me, but maybe there are other options. What do you think?

File: integration/container/logs_test.go [hunk 1]
```diff
@@ -32,3 +35,135 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+			)
+			defer client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{Force: true})
+
+			poll.WaitOn(t, container.IsStopped(ctx, client, id), poll.WithDelay(time.Millisecond*100))
+
+			logs, err := client.ContainerLogs(ctx, id, tC.logOps)
+			if logs != nil {
+				defer logs.Close()
+			}
+			assert.NilError(t, err)
```
[the remark this answers is not in the document]
Comment (corhere) [reply integration/container/logs_test.go]: Yes, that is clearer, and works nicely with the abort-on-failure semantics of `assert.NilError` since the test cannot meaningfully continue without a log reader. I like it!

File: daemon/logger/loggertest/logreader.go [hunk 2]
```diff
@@ -381,6 +383,75 @@ func (tr Reader) TestFollow(t *testing.T) {
...
+	go func() {
+		defer l.Close()
+		wg.Wait()
+	}()
+
+	lastStdout := 0
+	lastStderr := 0
+
+	// Check if the message count, order and content is equal to what was logged
+	for l := range lw.Msg {
```
Comment (corhere) [inline daemon/logger/loggertest/logreader.go:437]: Use the `readMessage` helper rather than looping directly through the `lw.Msg` channel as the helper handles timeouts and checking the `lw.Err` channel. (The `LogWatcher` API leaves much to be desired and could really use a refactor.) I suspect that the `msg="error closing current log file" error="invalid argument"` message you saw was accompanied by an error in the `lw.Err` channel but we don't know what the error was since the channel is never read from.
```suggestion
	for {
		l := readMessage(t, lw)
		if l == nil {
			break
		}

File: daemon/logger/loggertest/logreader.go [hunk 3]
```diff
@@ -381,6 +383,75 @@ func (tr Reader) TestFollow(t *testing.T) {
...
+		var counter *int
+		if l.Source == "stdout" {
+			counter = &lastStdout
+		} else if l.Source == "stderr" {
+			counter = &lastStderr
+		} else {
+			t.Fatalf("Corrupted message.Source = %s", l.Source)
```
Comment (corhere) [inline daemon/logger/loggertest/logreader.go:444]:
```suggestion
			t.Fatalf("Corrupted message.Source = %q", l.Source)
```

File: daemon/logger/jsonfilelog/read_test.go [hunk 4]
```diff
@@ -166,7 +166,7 @@ func TestFollowLogsWithRotation(t *testing.T) {
 					// set of logs that loggertest writes.
 					info.Config = map[string]string{
 						"compress": strconv.FormatBool(cmprs),
-						"max-size": "4096b",
+						"max-size": "512kb",
```
Comment (corhere) [inline daemon/logger/jsonfilelog/read_test.go:169]: This max-size is too big for `TestFollowLogsWithRotation` to test across a log-file rotation. Increasing it above 4096b invalidates the test.

**Comment by vvoland** [comment]
> Side note: that msg="error closing current log file" error="invalid argument" log message you saw when trying to test with follow doesn't look good. "invalid argument" is os.ErrInvalid, which is returned if you call ((*os.File)(nil)).Close(). That shouldn't happen. You might've found another bug! Is the current code in the PR branch sufficient to reproduce the failure?

Yes, this is reproducible on 8a4dcf5eb0a816bb29c26c4ed3e5a7c9721050d4 if you change `max-size` to small values. It never occurs without the missed rotations warning (from what I saw it was always more than 1 missed rotation).

The f is nil after this open:

https://github.com/vvoland/moby/blob/b34123ff3abcf763c4f4c5c4524a7f8969ad1c72/daemon/logger/loggerutils/follow.go#L74

err is:
```
open /var/folders/rv/7kmyxm9n6hd6xcpsh8tx3g300000gn/T/TestFollowLogsWithRotationcompress=falseConcurrent2959452999/001/logconcurrent0.log: no such file or directory
```

the return after `if err != nil` causes the defer block to be executed and we get the `((*os.File)(nil)).Close()`


So, I was writing this comment and debugging in the meantime. After some time I can no longer reproduce the above error, but now I get:
`time="2022-06-02T14:43:03+02:00" level=error msg="Error compressing log file after rotation" error="failed to open or create gzip log file: open /var/folders/rv/7kmyxm9n6hd6xcpsh8tx3g300000gn/T/TestFollowLogsWithRotationcompress=trueConcurrent4179564514/001/logconcurrent0.log.1.gz: no such file or directory"
`

This seems to be much rarer than the previous, but still it seems like the same situation, but another place?...

**Comment by vvoland** [comment]
Anyway, I applied your suggestions to the concurrent test. Thanks!

**Comment by vvoland** [comment]
Window tests fail for the tty case, thanks to the DeepEqual it is now clearly visible that the output also contains ANSI escape codes (which is probably expected in the TTY case).
I guess it is reasonable to strip those for comparing with expected output.
Might be a bit tricky, but I'll try with a simple removal of known sequences and see how that works.

File: integration/container/logs_test.go [hunk 5]
```diff
@@ -32,3 +38,145 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+			if tty && testEnv.OSType == "windows" {
+				stdoutStr = stripEscapeCodes(t, stdoutStr)
+			}
+			assert.DeepEqual(t, stdoutStr, tC.expectedOut)
+			assert.DeepEqual(t, stderr.String(), tC.expectedErr)
+		})
+	}
+}
+
+func stripEscapeCodes(t *testing.T, input string) string {
```
Comment (corhere) [inline integration/container/logs_test.go:162]: I found this module [github.com/Azure/go-ansiterm](https://pkg.go.dev/github.com/Azure/go-ansiterm#section-readme) which implements the full terminal-parsing state machine. (It's already vendored, too!) If you're feeling really keen, you could use it to make the test more robust by writing a "terminal emulator" fake which writes the printed characters to a string.

Here are some tips which might help if you want to give this a try:
- "C0 commands" are the non-printable ASCII codes including backspace `\b`, CR `\r`, LF `\n`
- You can embed an interface value into a struct to make the struct values implement the interface without explicitly defining all the interface methods. Any attempts to call these methods will result in a nil-dereference panic because it's literally calling a method on the embedded nil-valued interface field. (Unless you assign a concrete value to the field, of course.) I find it a very convenient trick when writing test fakes where only a subset of the interface is required for the test.

Comment (corhere) [reply integration/container/logs_test.go:162]: Never mind; turns out there's a bug in go-ansiterm's parsing of OSC commands (it incorrectly interprets a bare backslash character as the string terminator instead of `<ESC> \`) so it can't correctly strip out the Set Window Title command as it incorrectly transitions to the ground state at the backslash in `C:\`. https://go.dev/play/p/9xBvg_3_QKI

File: daemon/logger/loggertest/logreader.go [hunk 6]
```diff
@@ -381,6 +382,75 @@ func (tr Reader) TestFollow(t *testing.T) {
...
+			l.Log(copyLogMessage(m))
+		}
+	}
+	wg.Add(2)
+	go logAll(stdoutMessages)
+	go logAll(stderrMessages)
+	go func() {
+		defer l.Close()
+		wg.Wait()
+	}()
```
[the remark this answers is not in the document]
Comment (vvoland) [reply daemon/logger/loggertest/logreader.go:427]: Oh, good catch! I didn't see it, because I assumed that the `l.Close()` will cause the readMessage will return nil when the log is already closed.
Waiting for the l.Close to finish seems to fix the issue. Thanks!

File: integration/container/logs_test.go (hunk 5 above)
Comment (vvoland) [reply integration/container/logs_test.go:162]: Nice module and clever idea :) Would be good to use here if it didn't have the bug. However for now I'd go with the simple strings.Replace, since the same sequences are used most of the time. And if Windows still behaves funny, then I think we could skip the tests for Windows with TTY case and revisit this when your PR is accepted into the go-ansiterm.

File: integration/container/logs_test.go [hunk 7]
```diff
@@ -32,3 +38,148 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+	knownEscapes := []string{
+		"\x1b[2J",                        // Erase screen
+		"\x1b[m",                         // Reset color
+		"\x1b[Ht",                        // ??? The same as below but with trailing t
```
Comment (vvoland) [inline integration/container/logs_test.go:166]: Oh right, I get it. It moves to the beginning and writes `t` again (redundantly), so in the end output is correct (because our output starts with `this`)... But.. Why... 😿 

```
 === FAIL: github.com/docker/docker/integration/container TestLogs/driver_local/tty/stdout_and_stderr (2.96s)
    logs_test.go:178: Stripping: "\x1b[2J\x1b[?25l\x1b[m\x1b[Hthis is fine\b\x1b]0;C:\\bin\\sh.exe\x00\a\x1b[?25h\x1b[?25laccidents happen \b\x1b[?25h\x1b[Ht\x1b[1;29H"
    logs_test.go:156: assertion failed: 
        --- stdoutStr
        +++ tC.expectedOut
          string(
        - 	"his is fineaccidents happen",
        + 	"this is fineaccidents happen",
          )
        
        --- FAIL: TestLogs/driver_local/tty/stdout_and_stderr (2.96s)
```

**Review by vvoland — COMMENTED** [review]
Ok now I know why in this case `\x1b[Ht` still doesn't break the expected output.

File: integration/container/logs_test.go (hunk 5 above)
Comment (corhere) [reply integration/container/logs_test.go:166]: The go-ansiterm bug could feasibly be worked around by stripping the Set Window Title command sequence from the input before passing it to the parser. Since terminal commands are normally parsed character-by-character with a state machine, it must therefore be possible to write a regex which will match any possible Operating System Command sequence (which includes Set Window Title). It'd probably be less work than playing whack-a-mole with cursor movement commands.

Comment (corhere) [reply integration/container/logs_test.go:166]: I don't have permission to push to your branch, so I had to push to my own fork instead. Here you go: https://github.com/corhere/moby/commit/10614e994c4fcf37d2b4d76d1f9888084d09d917

File: integration/container/logs_test.go [hunk 8]
```diff
@@ -32,3 +38,133 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+	}{
+		// TTY, only one output streamś
+		{
+			desc: "tty/stdout and stderr",
+			tty:  true,
+			logOps: types.ContainerLogsOptions{
+				ShowStdout: true,
+				ShowStderr: true,
+			},
+			expectedOut: "this is fineaccidents happen",
```
[the remark this answers is not in the document]
Comment (vvoland) [reply integration/container/logs_test.go:72]: It's not really stripping, it's the backspace: `this is fine\b\x1b]0;C:\\bin\\sh.exe\x00\a`
I thought that maybe the `\b` is somehow cancelled out by the use of the `set title` + `\a` but I just checked it on my Windows 10 machine and this really rendered with the backspace.
However, running the test command on my Windows returns the correct output. But I'm not sure how my Windows version relates to those in our CI that have this issue.

[the remark this answers is not in the document]
Comment (corhere) [reply integration/container/logs_test.go:72]: > I thought that maybe the `\b` is somehow cancelled out by the use of the `set title` + `\a`

The `\a` character is one of two<sub>ish</sub> ways to terminate an OSC sequence such as `set title`. (The other 7-bit way is the two-byte "ST" sequence `0x1b 0x5c`, a.k.a. `ESC \`.) So technically, the `\a` is part of the `set title` command.

It looks like the weird backspacing only happens on Windows Server 2019. There are two distinct weird things in the logged string `this is fine\b\x1b]0;C:\\bin\\sh.exe\x00\a`: the backspace command before the OSC Set Window Title sequence, and the OSC command's argument is null-terminated when it shouldn't be. I booted up a Windows Server 2019 VM and played around a bit to characterize the issue. Note that I'm using Windows Powershell 5.1.
```
PS> docker build -t busybox --build-arg WINDOWS_BASE_IMAGE=mcr.microsoft.com/windows/servercore `
        --build-arg WINDOWS_BASE_IMAGE_TAG=ltsc2019 `
        .\contrib\busybox\
PS> docker run --rm -it busybox sh -c "echo -n this is fine; echo -n accidents happen >&2" | Format-Hex


           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

00000000   1B 5B 32 4A 1B 5B 3F 32 35 6C 1B 5B 6D 1B 5B 48  .[2J.[?25l.[m.[H
00000010   74 68 69 73 20 69 73 20 66 69 6E 65 08 1B 5D 30  this is fine..]0
00000020   3B 43 3A 5C 62 69 6E 5C 73 68 2E 65 78 65 00 07  ;C:\bin\sh.exe..
00000030   1B 5B 3F 32 35 68 1B 5B 3F 32 35 6C 61 63 63 69  .[?25h.[?25lacci
00000040   64 65 6E 74 73 20 68 61 70 70 65 6E 20 08 1B 5B  dents happen ..[
00000050   3F 32 35 68                                      ?25h

PS> docker run --rm -it busybox sh -c "echo -n this is fine; echo -n accidents happen >&2" | Write-Output
this is finaccidents happen
PS> docker run --rm -it busybox sh -c "echo -n this is fine; echo -n accidents happen >&2"
this is fineaccidents happen
```

AND IT GETS WEIRDER!

```
PS> docker run --name logweird1 -t busybox sh -c "echo -n this is fine; echo -n accidents happen >&2"
this is fineaccidents happen
PS> docker logs logweird1
this is fineaccidents happen
PS> get-content (docker inspect logweird1 --format '{{.LogPath}}')
{"log":"\u001b[2J\u001b[?25l\u001b[m\u001b[H\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8876173Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8876173Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8876173Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8876173Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8876173Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8882043Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8882043Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8907976Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\r\n","stream":"stdout","time":"2022-06-07T19:01:11.8915006Z"}
{"log":"\u001b[H\u001b]0;C:\\bin\\sh.exe\u0000\u0007\u001b[?25h\u001b[?25lthis is fineaccidents happen \u0008\u001b[?25h\u001b[Ht\u001b[1;29H","stream":"stdout","time":"2022-06-07T19:01:12.5417731Z"}
PS> docker run --name logweird2 -t busybox sh -c "echo -n this is fine; echo -n accidents happen >&2" | write-output
this is finaccidents happen
PS> get-content (docker inspect logweird2 --format '{{.LogPath}}')
{"log":"\u001b[2J\u001b[?25l\u001b[m\u001b[Hthis is fine\u0008\u001b]0;C:\\bin\\sh.exe\u0000\u0007\u001b[?25h\u001b[?25laccidents happen \u0008\u001b[?25h","stream":"stdout","time":"2022-06-07T19:04:06.1768156Z"}
```
Redirecting the `docker run` command's output **affects what gets logged by the daemon!** (`docker run -d` yields the same results as redirecting the output: erroneous backspace.) WHAT? HOW?? WHY??! 🤯 Since container logs always pass through the logging driver and the Engine API, there is no association between the CLI session and the container's pseudo-terminal which would be visible to the OS. The most likely explanation for this inconsistent behavior is that there is something different about the API calls the CLI makes when stdout is and is not a console which is necessary for the behavior to change. Curiously, I have found that writing a newline character to stdout first is sufficient to make redirected/detached output not backspace over the "e" in "fine."

The console-attached logs might hold clues to what's going wrong. The sequence of VT commands preceding the first printed characters decodes as follows:
* `ED(2)` -- erase entire display
* `DECTCEM(l)` -- hide cursor
* `SGR(0)` -- set all text formatting attributes to defaults
* `CUP(0;0)` -- move cursor to top-left
* `CR LF` * 49 -- move cursor down by 49 lines
* `CUP(0;0)` -- move cursor to top-left
* `OSC(0;"C:\\bin\\sh.exe\x00")` -- set window title to `C:\bin\sh.exe`
* `DECTCEM(h)` -- show cursor
* `DECTCEM(l)` -- hide cursor

My terminal window was sized to be 50 lines tall when I ran the command, so I don't think the 49 CRLFs is a coincidence. That sequence of commands looks like how a somewhat naive renderer would "repaint" a VT terminal to the state of a cleared text-mode screen buffer. This should come as no surprise as [rendering a Win32 Console API screen buffer to a stream of VT sequences is one of the jobs of conhost.exe](https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/#conhost---investing-in-yesterday-for-tomorrow). My theory is that perhaps the Windows Server 2019 conhost.exe (or the copy inside the base image?) VT renderer starts out in some bad state which results in backspace characters getting output in the wrong places and other rendering oddities, and that _setting the ConPTY's terminal dimensions_ has a side effect of putting the renderer into a good state. The Docker CLI keeps the container PTY dimensions synced with the attached terminal dimensions when stdout is a console, which would explain the differences in behavior between attached and detached/redirected. I have occasionally seen the "e" in "fine" overtyped when the output is not redirected, which could be explained by my theory as there being a race between the container outputting to the PTY and the Docker CLI setting the terminal dimensions. I have witnessed several variations of output by repeatedly running the same container multiple times so there's quite a bit of non-determinism in the PTY rendering, which either uninitialized memory in conhost.exe or a race condition could both potentially explain. Sleeping for half a second seems to result in deterministic output when outputting to an attached console, which is consistent with my set terminal dimensions race-condition theory.  But then again, I've also seen nondeterministic output on detached container runs that I couldn't reproduce when I added a sleep, so maybe there's more than one race in play?

In conclusion, the integration test found a bug with Windows containers, and conhost.exe might be involved.

[the remark this answers is not in the document]
Comment (vvoland) [reply integration/container/logs_test.go:72]: Thanks, that's a really impressive investigation :)
It's a bit strange that the resize actually does happen, since on Windows the container has the desired ConsoleSize already set on the creation. So there shouldn't be a resize at the start, unless you change your terminal size after `docker run`. But maybe under the hood there is still a delay between creating the container and setting the console size.
 
Anyway, I think we already spent a lot of time on those tests, so maybe for now we can skip this test on WS2019, report this to Microsoft or at least create an issue on our side, and revisit this once/if they handle it in a follow up?

[the remark this answers is not in the document]
Comment (thaJeztah) [reply integration/container/logs_test.go:72]: > What do we report to Microsoft?

Let me try some people from their side; @jstarks (I think you wrote most of the original go-ansiterm bits), or otherwise @dcanatah and @kevpar may have some ideas.



FWIW; I'm ok with `t.Skip()` some of these on Windows for now (with a link to this conversation of a tracking ticket), or perhaps we can do a limited "string contains `is fine` and /or string contains `accidents` check for windows (i.e., we know we received things on the right stream).

[the remark this answers is not in the document]
Comment (vvoland) [reply integration/container/logs_test.go:73]: > but it does look like in the docker run | write-output case [the console size configuration will be set to 0, 0](https://github.com/docker/cli/blob/c59773f1551a8fd289538efc82274332f31f8c19/cli/streams/out.go#L31-L35)

Yeah you are right, the size is 0,0 in this case, I missed the `GetTtySize`, so the resize does happen.

> What do we report to Microsoft?

I might have worded myself better - by reporting I mean "We encountered this issue, this doesn't happen on a newer Windows Server, so perhaps that means you fixed it and already know something more about it" rather than "plz fix it for us kthxbye".

> I don't see any container.WithConsoleSize options in your new logging integration tests. I wonder if adding it would change the test results.

Just tested, adding a non zero ConsoleSize does not change the result: https://github.com/moby/moby/runs/6812754016?check_suite_focus=true

BTW, I did test the Windows Server 2019 with the conhost.exe replaced (by modifying the image.. yeah that's really bad and evil but it works...). The conhost.exe from my WS2019 installation is the same as the one in the `ltsc2019` container (`10.0.17763.2746`). I replaced it with a conhost.exe from my PC's Windows 10 (`10.0.18362.1 `) and.. it works correctly! The backspace is no longer there and redirecting the stdout of `docker run` produces the same logfile without the additional CR LF (so same as `docker run ... | Write-Output` in your case).
I also tried the conhost.exe from Windows Server 2022 which is much newer (`10.0.20348.558`) and.. the terminal didn't start at all - probably due to other dependencies.

So, to conclude:
- conhost.exe is involved - it's sufficient to replace it to get fix the issue
- The relevant changes were made between 10.0.17763.2746 and 10.0.18362.1
- This still can be `conhost.exe happen to be better at clamping nonsensical configuration` though

File: integration/container/logs_test.go [hunk 9]
```diff
@@ -32,3 +39,151 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+				stdoutStr = stripEscapeCodes(t, stdoutStr)
+
+				// Special case on Windows Server 2019
+				// Check only that the output stream contains strings printed
+				// to stdout and stderr from inside container.
+				// This is a workaround for the backspace being output in an unexpected place
+				// which breaks the output: https://github.com/moby/moby/pull/43669#discussion_r891737623
+				if strings.Contains(testEnv.DaemonInfo.OperatingSystem, "Windows Server Version 1809") {
+					if tC.logOps.ShowStdout {
+						assert.Check(t, strings.Contains(stdout.String(), "this is fine"))
```
Comment (corhere) [inline integration/container/logs_test.go:165]: The string returned from `stripEscapeCodes` when the bug manifests is `"this is finaccidents happen"`. This check will return false and therefore the test will flake when that next happens, same as if it was not special-cased at all.

File: integration/container/logs_test.go [hunk 10]
```diff
@@ -32,3 +39,151 @@ func TestLogsFollowTailEmpty(t *testing.T) {
...
+						assert.Check(t, strings.Contains(stdout.String(), "accidents happen"))
```
Comment (corhere) [inline integration/container/logs_test.go:166]: `assert.Check`ing a boolean value without a custom failure message does not [yield a useful test failure](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures).
```
    logs_test.go:166: assertion failed: expression is false: strings.Contains(stdout.String(), "accidents happen"))
```
While you could write a custom failure message which interpolates the actual string, [gotest already has a rich comparison utility built-in which handles that for you.](https://pkg.go.dev/gotest.tools/v3@v3.1.0/assert/cmp#Contains)

**Review by corhere — CHANGES_REQUESTED** [review]

**Comment by vvoland** [comment]
Ok, so finally CI is green with the workaround for the Windows Server 2019.
Also sorry for the commit mess, I messed up the branches when squashing 🙈 

Extracted the comments regarding the Windows issue to #43710 - lets do a further investigation there

File: integration/container/logs_test.go (hunk 9 above)
Comment (vvoland) [reply integration/container/logs_test.go:165]: This compares the stdout.String() which is the raw unparsed string.

Comment (corhere) [reply integration/container/logs_test.go:165]: 🤦 I read the code to look for exactly that and still misread it. Oops, sorry about that.

Weren't there times when the initial `t` in `this` sometimes gets printed out of order? Would that cause the test to flake or was it just a redundant overtype which wouldn't confuse the test?

**Review by corhere — APPROVED** [review]

File: integration/container/logs_test.go (hunk 9 above)
Comment (vvoland) [reply integration/container/logs_test.go:165]: This was just the redundant overtype inserted at the end, but the correct string was still there:

\x1b[2J\x1b[?25l\x1b[m\x1b[H
**this is fine**
\b\x1b]0;C:\\bin\\sh.exe\x00\a\x1b[?25h\x1b[?25laccidents happen \b\x1b[?25h
**\x1b[Ht**
\x1b[1;29H

**Review by tianon — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!
